### PR TITLE
[Documentation] Add clarifying comment to `ClosestProof` documentation

### PR DESCRIPTION
**Pull request by h5law**
## Summary

Quick PR to add context to the closest proof method

## Issue

Fixes N/A

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all`
- [ ] **Run all/relevant benchmarks (if optimising)**: `make benchmark_{all | suite name}`

## Required Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code ([`godoc` format comments](https://go.dev/blog/godoc) see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))

### If Applicable Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated any relevant README(s)/documentation and left TODOs throughout the codebase
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

### CODE REVIEW

File: docs/SMT.md [hunk 1]
```diff
@@ -4,26 +4,26 @@
...
-  - [Closest Proof](#closest-proof)
-  - [Compression](#compression)
-  - [Serialisation](#serialisation)
+  * [Verification](#verification)
+  * [Closest Proof](#closest-proof)
+  * [Compression](#compression)
+  * [Serialisation](#serialisation)
 - [Database](#database)
-  - [Data Loss](#data-loss)
+  * [Data Loss](#data-loss)
```
Comment (bryanchriswhite) [inline docs/SMT.md:26]: Are these changes intentional?

File: docs/SMT.md [hunk 2]
```diff
@@ -378,10 +378,16 @@ This backstepping process allows the traversal to continue until it reaches a
...
-This method guarentees a proof of inclusion in all cases and can be verified by
+This method guarantees a proof of inclusion in all cases and can be verified by
 using the `VerifyClosestProof` function which requires the proof and root hash
 of the trie.
 
+NB: If the hash provided to the `ClosestProof` function is known prior to the
+tree being filled and closed there is the possibility of placing a leaf where
+the hash will lead. If used **as intended** the hash provided should **not** be
+known prior to calling the method and the tree should not be updateable after
+the fact.
```
Comment (bryanchriswhite) [inline docs/SMT.md:389]: > If the hash provided to the `ClosestProof` function is known prior to the
tree being filled and closed there is the possibility of placing a leaf where
the hash will lead.

Can you articulate what the consequences of this are?

**Review by bryanchriswhite — CHANGES_REQUESTED** [review]
:point_up_2:

File: docs/SMT.md (hunk 1 above)
Comment (h5law) [reply docs/SMT.md:26]: These are from my linter doing its thing

File: docs/SMT.md [hunk 3]
```diff
@@ -378,10 +378,21 @@ This backstepping process allows the traversal to continue until it reaches a
 sentinel leaf that has the longest common prefix and most bits in common with
 the provided hash, up to the depth of the leaf found.
 
-This method guarentees a proof of inclusion in all cases and can be verified by
+This method guarantees a proof of inclusion in all cases and can be verified by
 using the `VerifyClosestProof` function which requires the proof and root hash
 of the trie.
 
+NB: If the hash provided to the `ClosestProof` function is known prior to the
+tree being filled and closed there is the possibility of placing a leaf where
```
Comment (bryanchriswhite) [inline docs/SMT.md:386]:
```suggestion
tree being filled and closed, there is the possibility of a leaf being placed an attacker where
```

File: docs/SMT.md [hunk 4]
```diff
@@ -378,10 +378,21 @@ This backstepping process allows the traversal to continue until it reaches a
...
+the hash will lead. If used **as intended** the hash provided should **not** be
```
Comment (bryanchriswhite) [inline docs/SMT.md:387]:
```suggestion
the hash will lead. If used **as intended**, the hash provided should **not** be
```

File: docs/SMT.md [hunk 5]
```diff
@@ -378,10 +378,21 @@ This backstepping process allows the traversal to continue until it reaches a
...
+known prior to calling the method and the tree should not be updateable after
```
Comment (bryanchriswhite) [inline docs/SMT.md:388]:
```suggestion
known prior to calling the method; therefore, the tree should not be updateable after
```

File: docs/SMT.md [hunk 6]
```diff
@@ -378,10 +378,21 @@ This backstepping process allows the traversal to continue until it reaches a
...
+the fact. If the hash were known prior to closing the tree and not inserting
+into it anymore a leaf could be inserted into the tree in such a way that it
+would always be produced from the `ClosestProof` method (as it is deterministic).
```
Comment (bryanchriswhite) [inline docs/SMT.md:391]: Not sure if this is correct but I'm still struggling with this part. Can you rephrase "procuded from", I'm not sure what that means in this context.

```suggestion
the fact. If the last hash were known by <whom?>, after the last update but prior to closing, a leaf could be inserted into the tree in such a way that it would always be produced from the `ClosestProof` method (as it is deterministic).
```

**Review by bryanchriswhite — CHANGES_REQUESTED** [review]
Sorry for continuing to push on such a small thing but I think we're deriving more value by thinking it through.

File: docs/SMT.md (hunk 6 above)
Comment (h5law) [reply docs/SMT.md:391]: So essentially the caller of the method could place a leaf that would be guareneteed to be returned by the method IFF they know they hash prior to losing access to inserting into the tree. Have tried to make this clearer while keeping it general as this *could* be used in different ways from how we use it

File: docs/SMT.md [hunk 7]
```diff
@@ -378,10 +378,22 @@ This backstepping process allows the traversal to continue until it reaches a
...
+This method guarantees a proof of inclusion in all cases and can be verified by
 using the `VerifyClosestProof` function which requires the proof and root hash
 of the trie.
 
+NB: If the hash provided to the `ClosestProof` function is known prior to the
+tree being filled and closed there is the possibility of placing a leaf where
+the hash will lead. If used **as intended**, the hash provided should **not** be
+known prior to calling the method; therefore, the tree should not be updateable
+after the fact. If the hash were known by the caller of the method, after the
+last update but prior to closing the trie a leaf could be inserted into the tree
```
Comment (bryanchriswhite) [inline docs/SMT.md:390]: Are we calling it a tree or a trie?

File: docs/SMT.md [hunk 8]
```diff
@@ -378,10 +378,22 @@ This backstepping process allows the traversal to continue until it reaches a
...
 the provided hash, up to the depth of the leaf found.
 
-This method guarentees a proof of inclusion in all cases and can be verified by
+This method guarantees a proof of inclusion in all cases and can be verified by
 using the `VerifyClosestProof` function which requires the proof and root hash
 of the trie.
 
+NB: If the hash provided to the `ClosestProof` function is known prior to the
+tree being filled and closed there is the possibility of placing a leaf where
+the hash will lead. If used **as intended**, the hash provided should **not** be
```
Comment (bryanchriswhite) [inline docs/SMT.md:387]: I still am not clear on "who" we should be concerned about knowing the hash and "whom" would be placing a leaf where the hash will lead and whether they're the same actor.

File: docs/SMT.md (hunk 7 above)
Comment (h5law) [reply docs/SMT.md:390]: Technically it is a trie a prefix trie I renamed it a previous PR

**Review by bryanchriswhite — APPROVED** [review]
:heart:

**Comment by Olshansk** [comment]
@h5law This is a good addition and on a technical level contains all the details.

However, it's a lot of text & content to parse. For someone with context or who is 
willing to sit, think and read this many times, they'll understand. For 99% of OSS
contributors who will be too lazy (think about how many times we dive deep into other's
documentation), I think that bullet points + diagrams will make it easier to understand.

Even I feel like I need to read this multiple times and need to make notes along the way,
and the extra effort to make it easier to read will pay dividends down the road.

If you have time, I recommend approaching it like so.

```
## ClosestProof Use-Cases

The `CloestProof` function is intended for use in `commit & reveal` mechanism.

_NOTE: Throughout this document, `commitment` of the the tree's hashes is also referred to as closing the tree._

Take the following attack vector (**without** a commit prior to a reveal) into consideration:
1. The prover identifies the hash of a leaf they want to prove
2. The prover (prior to committing) inserts the hash ...
3. ...

Take the following normal flow (**with** a commit prior to reveal) as
1. ...
2. 
```
